### PR TITLE
Multi-threaded stress test (kernel mode)

### LINKED ISSUE #2097: Tests for race between program load / unload and NMR attach / detach
This issue is to track adding concurrency tests for the following cases:

Race | Test
-- | --
Program load (verify + JIT/Native) vs Program Info   provider attach/detach | Load/Unload an extension when a program is trying to   get loaded.
Program attach vs Hook provider attach/detach | Load/Unload an extension while Attach/Detach eBPF   program.
Program attach/detach vs Hook Invocation | Attach/Detach eBPF program while hook is firing events.

### PR DESCRIPTION
**Pull request by dv-msft**
## Description

This PR extends an existing stand-alone test suite by adding a multi-threaded stress tests for the in-kernel eBPF subsystem (`ebpfcore.sys, netwbpfext.sys`) itself.  The intent is to stress test the eBPF sub-system's resilience to multi-threaded stress scenarios. Details available in code comments.

These tests ensure deliberate races between the various user-mode initiated ebpf events such as `Load`, `Attach`, `Detach` and  `Close` etc., in a multi-threaded usage to simulate a badly written app's behavior and thus subject the eBPF kernel sub-system to races inherent in such multi-threaded scenarios.

Where applicable, the tests also optionally provide (via command line arguments) the ability to create an additional thread that keeps restarting the provider driver itself alongside the main test threads.

## Testing

This is new test collateral.

## Documentation

No additional documentation needed.

Fixes #2097 #1911 #2150 #2323

### CODE REVIEW

**Comment by dv-msft** [comment]
Converted back to draft to investigate ebpfcore bugs exposed by this test.

File: tests/sample/cgroup_sendto_count0.c
Comment (dv-msft) [inline tests/sample/cgroup_sendto_count0.c:1]: The ```cground_sendto_count<0-3>.c``` sample programs are being replaced with TCP 'connectv4/v6' variants instead to work around some existing UDP issues.  Will update the PR (and the associated test) shortly.

File: tests/sample/cgroup_count_connect4.c
Comment (dthaler) [inline tests/sample/cgroup_count_connect4.c:52]: What does "(simplifies user mode test implementation)" mean?
Compared to what?

File: tests/stress/ebpf_mt_stress.h
Comment (dthaler) [inline tests/stress/ebpf_mt_stress.h:65]: nit: most comments end in a period.  This one and line 71 don't.  Consistency would be good.

File: tests/stress/um/stress_tests_um.cpp
Comment (dthaler) [inline tests/stress/um/stress_tests_um.cpp:300]: a "thread ID" and an "interface index" are different things.  Should the "thread_index" field be renamed to "if_index"?

File: tests/stress/km/stress_tests_km.cpp
Comment (dthaler) [inline tests/stress/km/stress_tests_km.cpp:264]:
```suggestion
// is to ensure that such races do not cause hangs or crashes in the in-kernel eBPF subsystem components'
```

File: tests/stress/km/stress_tests_km.cpp
Comment (dthaler) [inline tests/stress/km/stress_tests_km.cpp:1028]:
```suggestion
    //        the count (TCP) 'connect' attempts in the 'connect4_count_map' map at its port.
```

**Review by dthaler — COMMENTED** [review]
Looks good to me, other than some editorial nits.

File: tests/sample/cgroup_count_connect4.c
Comment (dv-msft) [reply tests/sample/cgroup_count_connect4.c]: The (misguided) attempt at brevity made it misleading :-P Expanded the comment to provide the correct rationale.

File: tests/sample/cgroup_count_connect6.c
Comment (dv-msft) [reply tests/sample/cgroup_count_connect6.c]: As before, the (misguided) attempt at brevity made it misleading :-P Expanded the comment to provide the correct rationale.

File: tests/stress/ebpf_mt_stress.h
Comment (dv-msft) [reply tests/stress/ebpf_mt_stress.h]: fixed

File: tests/stress/um/stress_tests_um.cpp
Comment (dv-msft) [reply tests/stress/um/stress_tests_um.cpp:299]: The ```thread_index``` member really is really meant to be used an index into the more-than-one arrays that the threads need to access and is used accordingly.  This value is guaranteed to be present in each thread context _and_ be unique.  This fit the bill for the if_index value requirements and I used it accordingly.

I can certainly add in a new ```if_index``` field in the context, or add a comment explaining the (overloaded) usage here.

File: tests/stress/km/stress_tests_km.cpp
Comment (dv-msft) [reply tests/stress/km/stress_tests_km.cpp]: This was me going overboard in my std::format usage :-D In hindsight, this specific use case is too trivial (compared to the boilerplate code required for the custom formatter implementation) so yanked it.

File: tests/stress/km/stress_tests_km.cpp
Comment (dv-msft) [reply tests/stress/km/stress_tests_km.cpp]: fixed

File: tests/stress/km/stress_tests_km.cpp
Comment (dv-msft) [reply tests/stress/km/stress_tests_km.cpp]: fixed

File: tests/stress/km/stress_tests_km.cpp
Comment (dv-msft) [reply tests/stress/km/stress_tests_km.cpp]: fixed

File: tests/stress/km/stress_tests_km.cpp
Comment (dv-msft) [reply tests/stress/km/stress_tests_km.cpp]: I will be filing a new (backlog) issue to look into this.   For now, I've removed the dead code.

File: tests/stress/ebpf_mt_stress.h
Comment (dv-msft) [reply tests/stress/ebpf_mt_stress.h]: fixed.

File: tests/stress/um/stress_tests_um.cpp
Comment (dv-msft) [reply tests/stress/um/stress_tests_um.cpp:303]: Added a comment explaining the overloaded usage.

File: tests/stress/km/stress_tests_km.cpp
Comment (dv-msft) [reply tests/stress/km/stress_tests_km.cpp]: fixed

File: tests/stress/km/stress_tests_km.cpp
Comment (dv-msft) [reply tests/stress/km/stress_tests_km.cpp]: fixed

File: tests/stress/um/stress_tests_um.cpp
Comment (dthaler) [inline tests/stress/um/stress_tests_um.cpp:17]: Can remove nullptr check to reduce code complexity.

File: tests/stress/um/stress_tests_um.cpp
Comment (dthaler) [inline tests/stress/um/stress_tests_um.cpp:167]:
```suggestion
        uint32_t if_index = test_params.if_index;
```

**Review by dthaler — DISMISSED** [review]

**Review by Alan-Jowett — DISMISSED** [review]

File: tests/stress/readme.md
Comment (dv-msft) [reply tests/stress/readme.md:59]: fixed

File: tests/stress/um/stress_tests_um.cpp
Comment (dv-msft) [reply tests/stress/um/stress_tests_um.cpp:167]: added comment.

File: tests/stress/um/stress_tests_um.cpp
Comment (dv-msft) [reply tests/stress/um/stress_tests_um.cpp:17]: fixed

**Review by dthaler — DISMISSED** [review]

**Comment by dv-msft** [comment]
Blocked on [PR](https://github.com/microsoft/ebpf-for-windows/pull/2340) for CI/CD FuzzerDebug related build break.  Will rebase and update once that PR is merged.

File: .github/workflows/cicd.yml
Comment (dthaler) [inline .github/workflows/cicd.yml:416]: Since you're adding reusable-test.yml instances in cicd.yml, I think you need to update the after_n_builds count in codecov.yml

File: tests/stress/km/stress_tests_km.cpp
Comment (dthaler) [inline tests/stress/km/stress_tests_km.cpp:27]:
```suggestion
// or kernel mode test suites. (For example, some programs could be meant for kernel mode stress testing only.)
```

File: .github/workflows/cicd.yml
Comment (dv-msft) [reply .github/workflows/cicd.yml:416]: As checked in, `.github\codecov.yml` has the `after_n_builds` field set to 23 which already exceeds the new actual count (22) _including_ the new tests added in this PR:

```
D:\wrk\ebpf-for-windows\.github [mt-stress-tests-km]
PS > findstr after_n_builds .\codecov.yml
    after_n_builds: 23 # (reusable-test.yml instances in cicd.yml) * (# configurations per instance)
    after_n_builds: 23 # (reusable-test.yml instances in cicd.yml) * (# configurations per instance)


D:\wrk\ebpf-for-windows\.github [mt-stress-tests-km]
PS > findstr uses .\workflows\cicd.yml | findstr reusable-test.yml | measure-object -line


Lines Words Characters Property
----- ----- ---------- --------
   22
```

What am I missing?

File: tests/stress/km/stress_tests_km.cpp
Comment (dv-msft) [reply tests/stress/km/stress_tests_km.cpp]: fixed

File: .github/workflows/cicd.yml
Comment (dthaler) [reply .github/workflows/cicd.yml:416]: You're missing the `* (# configurations per instance)` part.

Some of those uses have 2 configurations (Debug, Release) each, some have 1 (FuzzerDebug).

Comment (dv-msft) [reply .github/workflows/cicd.yml:416]: Thanks, Dave. That clears it up.  Now I'm waffling on whether we need to run MT stress on debug _and/or_ release builds as I can argue for both sides :-P (Personal preference being 'release only'.)

I'm not aware of a similar precedence here (or quite possibly overlooking an existent instance) so I'll go with the team consensus. Maybe we could discuss this during tomorrow's meeting?

Comment (dv-msft) [reply .github/workflows/cicd.yml:416]: The current value for `after_n_builds` counter in `.github\codecov.yml` is incorrect.  I've filed an [issue](https://github.com/microsoft/ebpf-for-windows/issues/2373) to address this.

Also, only tasks that upload code coverage reports need to be accounted for in that counter and the tasks added in this PR do not do so.  As a result, this PR does not need to update the `after_n_builds` counter.

**Review by dthaler — APPROVED** [review]

**Review by Alan-Jowett — APPROVED** [review]

**Pull request merged**
